plutus-ir/src/Language/PlutusIR/Transform/LetFloat.hs

```diff
@@ -0,0 +1,455 @@
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE LambdaCase       #-}
+module Language.PlutusIR.Transform.LetFloat (floatTerm) where
+
+import           Language.PlutusIR
+import           Language.PlutusIR.Analysis.Dependencies
+import           Language.PlutusIR.Analysis.Free
+
+import           Control.Lens
+import           Control.Monad.Reader
+import           Control.Monad.RWS
+import           Control.Monad.State
+
+import qualified Language.PlutusCore                     as PLC
+import qualified Language.PlutusCore.Name                as PLC
+
+import qualified Algebra.Graph.AdjacencyMap              as AM
+import qualified Algebra.Graph.AdjacencyMap.Algorithm    as AM
+import qualified Algebra.Graph.NonEmpty.AdjacencyMap     as AMN
+
+import           Data.Foldable                           (fold)
+import           Data.Function                           (on)
+import qualified Data.IntMap                             as IM
+import qualified Data.Map                                as M
+import           Data.Maybe                              (fromMaybe)
+import qualified Data.Set                                as S
+
+{- Note: [Float algorithm]
+
+Our algorithm is influenced by the algorithm described in Section-4 of:
+Peyton Jones, Simon, Will Partain, and Andre Santos. "Let-Floating: Moving Bindings to Give Faster Programs." In Proceedings of the First ACM SIGPLAN International Conference on Functional Programming, 1-12. ICFP '96. New York, NY, USA: ACM, 1996. https://doi.org/10.1145/232627.232630.
+
+Our algorithm tries to do two things:
+
+a) move let-bindings as outwards as possible (a.k.a. full laziness).
+b) merge any floated,let-NonRec bindings which are adjacent, into a multi-let group (for readability purposes).
+
+Our algorithm is comprised of two passes of the PlutusIR AST:
+
+- 1st pass: Compute the "maximum ranks" of all lets in a given term. A maximum rank signifies a source-code position that the let will be moved/floated to during the 2nd pass.
+- 2nd pass: Move lets to their maximum rank positions. The pass makes sure that same-rank lets are grouped in multi-let-groups which
+are topologically sorted by their dependency graph (see 'Dependencies.runTermDeps').
+
+-}
+
+{- Note: [Float first-pass]
+
+Visits a PIR term and marks all identifiers introduced by lets with their "maximum rank" (level number in the paper).
+
+A maximum rank of a let-bound identifier can be seen as its "deepest" (lambda,Lamdba,or let)-dependency that occurs free in that let's RHS.
```

Comment:
It's worth specifying somewhere that a "dependency" is just a bound variable (I was confused by this term).

Reply:
Also, as per discussion in slack, this needs a few examples.